controller/src/main/java/io/pravega/controller/server/ControllerService.java

```diff
@@ -328,10 +312,10 @@ private SegmentRange convert(final String scope,
                     return new ImmutablePair<>(data.getId(), getSegmentRanges(segments, scope, stream));
                 }).handle((result, ex) -> {
                     if (ex != null) {
-                        transactionMetrics.createTransactionFailed(scope, stream);
+                        TransactionMetrics.getInstance().createTransactionFailed(scope, stream);
```

Comment:
do we need to call getInstance each time? Can we call it once and store it in a field in this object and use it for the lifecycle of this object?

Reply:
Calling getInstance which is synchronized from so many place, and from potentially multiple threads for different workflows, will result in lot of lock contention and waits which could be detrimental for performance. 
In the constructor we can call it once and set it in a local field and use it throughout this class.

Reply:
As far as I know, this is how the traditional singleton pattern defines to get the instance. However, if you want, I could "embed" the `getInstance()` call on each method of `TransactionMetrics` and `StreamMetrics`, so from the usage perspective, it would be virtually like calling a static method (i.e., no need to explicitly call `getInstance()`). Is that better? I think that what we want to avoid is to add extra fields to classes just to report these metrics, so created a reference field on each class would be kind of against the change, right?

Reply:
The concern is everytime you call getInstance you contest for lock. 

Instead, you can call getInstance once when an object of this class was created  -- e.g. ControllerServiceStarter's constructor and keep that reference in a field in ControllerServiceStarter.

So any subsequent call in ControllerServiceStarter for metrics will use the referenced object instead of calling getInstance again.